lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs

```diff
@@ -0,0 +1,454 @@
+-- random) which means, any user of cardano-wallet.
+emptyGenesis :: W.BlockchainParameters -> W.Block
+emptyGenesis bp = W.Block
+    { transactions = []
+    , delegations  = []
+    , header = W.BlockHeader
+        { slotId =
+            W.SlotId 0 0
+        , blockHeight =
+            Quantity 0
+        , headerHash =
+            coerce $ W.getGenesisBlockHash bp
+        , parentHeaderHash =
+            W.Hash (BS.replicate 32 0)
+        }
+    }
+
+--------------------------------------------------------------------------------
+--
+-- Genesis
+
+
+genesisTip :: Tip (ShelleyBlock c)
+genesisTip = legacyTip genesisPoint genesisBlockNo
+  where
+    -- NOTE: ourobouros-network states that:
+    --
+    -- There /is/ no block number if we are at genesis
+    -- ('genesisBlockNo' is the block number of the first block on the chain).
+    -- Usage of this function should be phased out.
+    genesisBlockNo = BlockNo 0
+
+
+--------------------------------------------------------------------------------
+--
+-- Network Parameters
+
+-- | Settings for configuring a MainNet network client
+mainnetVersionData
+    :: NodeVersionData
+mainnetVersionData =
+    ( NodeToClientVersionData
+        { networkMagic =
+            NetworkMagic $ fromIntegral $ W.getProtocolMagic W.mainnetMagic
+        }
+    , nodeToClientCodecCBORTerm
+    )
+
+-- | Settings for configuring a TestNet network client
+testnetVersionData
+    :: W.ProtocolMagic
+    -> NodeVersionData
+testnetVersionData pm =
+    ( NodeToClientVersionData
+        { networkMagic =
+            NetworkMagic $ fromIntegral $ W.getProtocolMagic pm
+        }
+    , nodeToClientCodecCBORTerm
+    )
+
+--------------------------------------------------------------------------------
+--
+-- Type Conversions
+
+-- fixme: maybe just toShelleyHash = ShelleyHash . CC.unsafeHashFromBytes
+toShelleyHash :: Crypto c => W.Hash "BlockHeader" -> (ShelleyHash c)
+toShelleyHash (W.Hash bytes) =
+    unsafeDeserialiseCbor fromCBOR
+    $ BL.fromStrict bytes
+
+toEpochSlots :: W.EpochLength -> EpochSlots
+toEpochSlots =
+    EpochSlots . fromIntegral . W.unEpochLength
+
+-- | Magic value for the absence of a block.
+hashOfNoParent :: W.Hash "BlockHeader"
+hashOfNoParent = W.Hash . BS.pack $ replicate 0 32
+
+toPoint
+    :: Crypto c
+    => W.Hash "Genesis"
+    -> W.EpochLength
+    -> W.BlockHeader
+    -> Point (ShelleyBlock c)
+toPoint genesisH epLength (W.BlockHeader sid _ h _)
+  | h == (coerce genesisH) = O.GenesisPoint
+  | otherwise = O.Point $ Point.block (toSlotNo epLength sid) (toShelleyHash h)
+
+toSlotNo :: W.EpochLength -> W.SlotId -> SlotNo
+toSlotNo epLength =
+    SlotNo . W.flatSlot epLength
+
+-- | SealedTx are the result of rightfully constructed byron transactions so, it
+-- is relatively safe to unserialize them from CBOR.
+toGenTx :: HasCallStack => W.SealedTx -> GenTx (ShelleyBlock c)
+toGenTx = error "todo: toGenTx not implemented"
+
+fromShelleyTx :: SL.Tx c -> W.Tx
```

Comment:
minor note: the conventions I have been trying to enforce in this file were that:

a) Identifiers from cardano-wallet are used qualified.
b) Identifiers from networking, consensus or ledger were used unqualified.  
c) Function are named after the data-type
d) `toXXX` when they translate from a cardano-wallet type to a networking, consensus & ledger data-type. `fromXXX` when they do the opposite (from networking to cardano-wallet).

Reply:
ouroboros-network seems to consistently imports everything from the Shelley ledger as qualified `as SL`.

There are name clashes (e.g. `Block`, `Tx`) between ouroboros-network and the specs.

I think we have to use `qualified` to some extent. At least for `SL` and `W`.